stdlib/random.ml

```diff
@@ -114,35 +114,107 @@ module State = struct
+     [bound] may be any positive [int].  [mask] must be of the form [2{^i}-1]
+     and greater or equal to [n].  Larger values of [mask] make the function
+     run faster (fewer samples are rejected).  Smaller values of [mask]
+     are usable on a wider range of OCaml implementations.  *)
+  let rec int_aux s n mask =
+    (* We start by drawing a non-negative integer in the [ [0, mask] ] range *)
+    let r = Int64.to_int (next s) land mask in
     let v = r mod n in
-    if r - v > 0x3FFFFFFF - n + 1 then intaux s n else v
+    (* For uniform distribution of the result between 0 included and [n]
+     * excluded, the random number [r] must have been drawn uniformly in
+     * an interval whose length is a multiple of [n]. To achieve this,
+     * we use rejection sampling on the greatest interval [ [0, k*n-1] ]
+     * that fits in [ [0, mask] ].  That is, we reject the
+     * sample if it falls outside of this interval, and draw again.
+     * This is what the test below does, while carefuly avoiding
+     * overflows and sparing a division [mask / n]. *)
+    if r - v > mask - n + 1 then int_aux s n mask else v
 
+  (* Return an integer between 0 (included) and [bound] (excluded).
+     The bound must fit in 31-bit signed integers.
+     This function yields the same output regardless of the integer size. *)
   let int s bound =
-    if bound > 0x3FFFFFFF || bound <= 0
+    if bound > max_int31 || bound <= 0
     then invalid_arg "Random.int"
-    else intaux s bound
+    else int_aux s bound max_int31
 
   (* Return an integer between 0 (included) and [bound] (excluded).
      [bound] may be any positive [int]. *)
-  let rec int63aux s n =
-    let r = Int64.to_int (next s) land max_int in
-    let v = r mod n in
-    if r - v > max_int - n + 1 then int63aux s n else v
-
   let full_int s bound =
     if bound <= 0 then
       invalid_arg "Random.full_int"
-    else if bound > 0x3FFFFFFF then
-      int63aux s bound
+    (* When the bound fits in 31-bit signed integers, we use the same mask
+       as in function [int] so as to yield the same output on all platforms
+       supported by OCaml (32-bit OCaml, 64-bit OCaml, and JavaScript).
+       When the bound fits in 32-bit signed integers, we use [max_int32]
+       as the mask so as to yield the same output on all platforms where
+       [Sys.int_size >= 32] (i.e. JavaScript and 64-bit OCaml). *)
     else
-      intaux s bound
+      int_aux s bound
+        (if bound <= max_int31 then max_int31
+         else if bound <= max_int32 then max_int32
+         else max_int)
+
+  (* Return an integer between [min] (included) and [max] (included).
+     The [nbits] parameter is the size in bits of the signed integers
+     we draw from [s].
+     We must have [-2{^nbits - 1} <= min <= max < 2{^nbits - 1}].
+     Moreover, for the iteration to converge quickly, the interval
+     [[min, max]] should have width at least [2{^nbits - 1}]. *)
+  let rec int_in_range_alt s ~min ~max ~nbits =
```

Comment:
Should we document that near the `2^{nbits - 1}` limit, we have a quite high standard deviation in the number of samples (around `2+ε`)?
The suffix `_alt` is also quite unclear, and it is not reused for the other variants of those functions: Maybe `int_in_large_range` would be more readable?

Reply:
OK, done (7ebfbe4f89).